chain/client/src/sync/epoch.rs

```diff
@@ -514,6 +508,148 @@ impl EpochSync {
 
         Ok(())
     }
+
+    // Verify EpochSyncProof
+    fn verify_proof(
+        &self,
+        proof: &EpochSyncProof,
+        epoch_manager: &dyn EpochManagerAdapter,
+    ) -> Result<(), Error> {
+        let EpochSyncProof { past_epochs, last_epoch, current_epoch } = proof;
+        if past_epochs.is_empty() {
+            return Err(Error::InvalidEpochSyncProof("empty past_epochs".to_string()));
+        }
+
+        // Verify block producer handoff to the first epoch after genesis
+        Self::verify_block_producer_handoff(
+            past_epochs.first().unwrap(),
+            self.genesis.epoch_id(),
+            self.genesis.next_bp_hash(),
+            epoch_manager,
+        )?;
+
+        // Verify block producer handoff between all past epochs
+        for epoch_index in 1..past_epochs.len() {
+            let epoch = &past_epochs[epoch_index];
+            let prev_epoch = &past_epochs[epoch_index - 1];
+            Self::verify_block_producer_handoff(
+                epoch,
+                prev_epoch.last_final_block_header.epoch_id(),
+                prev_epoch.last_final_block_header.next_bp_hash(),
+                epoch_manager,
+            )?;
+        }
+
+        Self::verify_epoch_sync_data_hash(
+            &last_epoch,
+            &past_epochs.last().unwrap().block_producers,
+            epoch_manager,
+        )?;
+
+        Self::verify_current_epoch_data(
+            current_epoch,
+            &last_epoch.final_block_header_in_next_epoch,
+        )?;
+        Ok(())
+    }
+
+    fn verify_current_epoch_data(
+        current_epoch: &EpochSyncProofCurrentEpochData,
+        final_block_header: &BlockHeader,
+    ) -> Result<(), Error> {
+        // Verify first_block_header_in_epoch
+        let first_block_header = &current_epoch.first_block_header_in_epoch;
+        if !verify_hash(
```

Comment:
nit: rename it verify_block_merkle_proof, or explicitly mention the module name here, for readability? Without the context of "merkle", `verify_hash` doesn't mean anything.